File: integrations/stitch/stitch_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
🎨 AMB_V2 - Google Stitch SDK Client & Tooling (SRP)
Localização: amb_v2/integrations/stitch/stitch_client.py
Responsabilidade Única: Interface Python unificada para geração de telas, refinamentos,
variantes visuais e sincronização de Design Tokens via Stitch Runner Node.js.
"""

import os
import sys
import json
import subprocess
from typing import Dict, Any, Optional

from config import Colors, log, log_error, require_env, get_env, find_repo_root, ApiExecutionError


class StitchClient:
    """Client oficial para o Google Stitch SDK."""

    def __init__(self, project_id: Optional[str] = None):
        self.api_key = require_env("STITCH_API_KEY")
        self.project_id = project_id or get_env("STITCH_PROJECT_ID")
        self.runner_path = self._resolve_runner()

    def _resolve_runner(self) -> str:
        candidates = [
            os.path.abspath(os.path.join(os.path.dirname(__file__), "stitch_client.mjs")),
            os.path.abspath(os.path.join(os.path.dirname(__file__), "client", "stitch_client.mjs")),
        ]
        runner = next((p for p in candidates if os.path.exists(p)), None)
        if not runner:
            raise ApiExecutionError(f"Runner Stitch (stitch_client.mjs) não encontrado em {candidates[0]}")
        return runner

    def _run_node_command(self, action: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Executa comando no runner Node.js do Stitch SDK."""
        proc = subprocess.run(
            ["node", self.runner_path, action, json.dumps(payload)],
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace"
        )
        if proc.returncode != 0:
            err_msg = proc.stderr.strip() or proc.stdout.strip()
            try:
                err_json = json.loads(err_msg)
                err_msg = err_json.get("error", err_msg)
            except Exception:
                pass
            raise ApiExecutionError(f"Falha na execução do Stitch SDK ({action}): {err_msg}")

        try:
            return json.loads(proc.stdout.strip())
        except json.JSONDecodeError:
            return {"output": proc.stdout.strip()}

    def generate_screen(
        self,
        prompt: str,
        device_type: str = "DESKTOP",
        model_id: str = "GEMINI_3_1_PRO",
        design_system: Optional[str] = None
    ) -> Dict[str, Any]:
        """Gera uma nova tela visual a partir de uma descrição textual."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        payload = {
            "projectId": project_id,
            "prompt": prompt,
            "deviceType": device_type,
            "modelId": model_id,
        }
        if design_system:
            payload["designSystem"] = design_system

        log("STITCH", f"Disparando geração de nova tela visual ({device_type})...", Colors.CYAN)
        res = self._run_node_command("generate_screen", payload)
        return res

    def edit_screen(self, screen_id: str, prompt: str) -> Dict[str, Any]:
        """Refina e edita uma tela existente com novas instruções visuais."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        payload = {
            "projectId": project_id,
            "screenId": screen_id,
            "prompt": prompt
        }
        log("STITCH", f"Refinando tela {screen_id}...", Colors.CYAN)
        res = self._run_node_command("edit_screen", payload)
        return res

    def get_screen(self, screen_id: str) -> Dict[str, Any]:
        """Obtém detalhes, DOM HTML e screenshot de uma tela."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        payload = {
            "projectId": project_id,
            "screenId": screen_id
        }
        log("STITCH", f"Consultando detalhes da tela {screen_id}...", Colors.CYAN)
        res = self._run_node_command("get_screen", payload)
        return res

    def generate_variants(self, screen_id: str, prompt: str = "Explorar variações visuais", variant_count: int = 3) -> Dict[str, Any]:
        """Gera variantes visuais exploratórias a partir de uma tela."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        payload = {
            "projectId": project_id,
            "selectedScreenIds": [screen_id],
            "prompt": prompt,
            "variantOptions": {
                "variantCount": variant_count,
                "creativeRange": "EXPLORE"
            }
        }
        log("STITCH", f"Gerando {variant_count} variantes para a tela {screen_id}...", Colors.CYAN)
        res = self._run_node_command("generate_variants", payload)
        return res

    def sync_design_system(self, design_md_path: Optional[str] = None) -> Dict[str, Any]:
        """Sincroniza arquivo design.md com o Design System do Stitch."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        root = find_repo_root()
        default_paths = [
            os.path.join(root, "design.md"),
            os.path.join(root, "docs", "design.md"),
            os.path.join(root, ".antigravity", "rules", "design.md"),
        ]
        resolved_file = design_md_path or next((p for p in default_paths if os.path.exists(p)), None)
        if not resolved_file or not os.path.exists(resolved_file):
            raise ApiExecutionError("Arquivo design.md não encontrado para sincronização de Design Tokens.")

        with open(resolved_file, "r", encoding="utf-8") as f:
            md_content = f.read()

        payload = {
            "projectId": project_id,
            "designMd": md_content
        }
        log("STITCH-DS", f"Sincronizando {resolved_file} com o projeto {project_id}...", Colors.CYAN)
        res = self._run_node_command("create_design_system_from_design_md", payload)
        return res


# Funções utilitárias avulsas para import direto
def generate_screen(prompt: str, **kwargs) -> Dict[str, Any]:
    return StitchClient().generate_screen(prompt, **kwargs)

def edit_screen(screen_id: str, prompt: str) -> Dict[str, Any]:
    return StitchClient().edit_screen(screen_id, prompt)

def get_screen(screen_id: str) -> Dict[str, Any]:
    return StitchClient().get_screen(screen_id)

def generate_variants(screen_id: str, prompt: str = "Explorar variações", count: int = 3) -> Dict[str, Any]:
    return StitchClient().generate_variants(screen_id, prompt, variant_count=count)

def sync_design_system(design_md_path: Optional[str] = None) -> Dict[str, Any]:
    return StitchClient().sync_design_system(design_md_path)

File: integrations/stitch/stitch_client.mjs
#!/usr/bin/env node
/**
 * 🎨 AMB_V2 - Runner Node.js do Google Stitch SDK (@google/stitch-sdk)
 * Localização: amb_v2/integrations/stitch/stitch_client.mjs
 * Responsabilidade Única: Estabelecer conexão com o StitchToolClient e despachar chamadas oficiais.
 */

import { StitchToolClient, stitch } from '@google/stitch-sdk';
import fs from 'fs';
import path from 'path';

function getApiKey() {
  if (process.env.STITCH_API_KEY) return process.env.STITCH_API_KEY;
  if (process.env.GEMINI_API_KEY) return process.env.GEMINI_API_KEY;

  let curr = process.cwd();
  for (let i = 0; i < 6; i++) {
    const envFile = path.join(curr, '.env');
    if (fs.existsSync(envFile)) {
      const txt = fs.readFileSync(envFile, 'utf-8');
      for (const line of txt.split('\n')) {
        const t = line.trim();
        if (t.startsWith('STITCH_API_KEY=')) {
          return t.split('=')[1].trim().replace(/^['"]|['"]$/g, '');
        }
      }
    }
    const parent = path.dirname(curr);
    if (parent === curr) break;
    curr = parent;
  }
  return null;
}

async function extractScreenDetails(client, projectId, result) {
  let screenId = null;
  let screenshotUrl = null;
  let htmlCode = null;
  let title = "";
  let description = "";
  const messages = [];

  if (result && typeof result === 'object') {
    if (result.screenId) screenId = result.screenId;
    if (result.name && typeof result.name === 'string') screenId = result.name.split('/').pop();
    if (result.screenshot?.downloadUrl) screenshotUrl = result.screenshot.downloadUrl;
    if (result.screenshotUrl) screenshotUrl = result.screenshotUrl;
    if (result.htmlCode?.downloadUrl) htmlCode = result.htmlCode.downloadUrl;
    if (typeof result.htmlCode === 'string') htmlCode = result.htmlCode;
    if (result.title) title = result.title;
    if (result.description) description = result.description;
    if (result.message) messages.push(result.message);

    const comps = result.outputComponents || result.output_components || [];
    for (const comp of comps) {
      if (comp.text) messages.push(comp.text.trim());
      if (comp.design?.screens) {
        for (const s of comp.design.screens) {
          if (s.name) screenId = s.name.split('/').pop();
          if (s.screenshot?.downloadUrl) screenshotUrl = s.screenshot.downloadUrl;
          if (s.htmlCode?.downloadUrl) htmlCode = s.htmlCode.downloadUrl;
          if (s.title) title = s.title;
          if (s.description) description = s.description;
        }
      }
    }
  }

  if (htmlCode && typeof htmlCode === 'string' && (htmlCode.startsWith('http://') || htmlCode.startsWith('https://'))) {
    try {
      const resp = await fetch(htmlCode);
      if (resp.ok) {
        htmlCode = await resp.text();
      }
    } catch {}
  }

  return {
    success: true,
    screenId,
    screenshotUrl,
    htmlCode,
    title,
    description,
    messages
  };
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length < 2) {
    console.error(JSON.stringify({
      error: "Argumentos insuficientes.",
      usage: "node stitch_client.mjs <command> '<json_payload>'"
    }));
    process.exit(1);
  }

  const [command, rawPayload] = args;
  let payload = {};
  try {
    payload = JSON.parse(rawPayload);
  } catch (e) {
    console.error(JSON.stringify({ error: `Payload JSON inválido: ${e.message}` }));
    process.exit(1);
  }

  const apiKey = getApiKey();
  if (!apiKey) {
    console.error(JSON.stringify({
      error: "STITCH_API_KEY ausente.",
      hint: "Configure STITCH_API_KEY no arquivo .env"
    }));
    process.exit(1);
  }

  try {
    const client = new StitchToolClient({ apiKey });
    let response;

    switch (command) {
      case "generate_screen":
      case "generate_screen_from_text": {
        const res = await client.callTool("generate_screen_from_text", payload);
        response = await extractScreenDetails(client, payload.projectId, res);
        break;
      }
      case "generate_variants": {
        const res = await client.callTool("generate_variants", payload);
        response = await extractScreenDetails(client, payload.projectId, res);
        break;
      }
      case "edit_screens":
      case "edit_screen": {
        const res = await client.callTool("edit_screens", payload);
        response = await extractScreenDetails(client, payload.projectId, res);
        break;
      }
      case "get_screen": {
        const res = await client.callTool("get_screen", {
          projectId: payload.projectId,
          screenId: payload.screenId,
          name: `projects/${payload.projectId}/screens/${payload.screenId}`
        });
        response = await extractScreenDetails(client, payload.projectId, res);
        try {
          const projResp = await client.callTool("get_project", { name: `projects/${payload.projectId}` });
          const proj = projResp?.project || projResp;
          if (proj?.designTheme?.designMd) {
            response.designMd = proj.designTheme.designMd;
          }
        } catch {}
        break;
      }
      case "get_project": {
        const projName = payload.name || (payload.projectId ? `projects/${payload.projectId}` : (payload.id ? `projects/${payload.id}` : ''));
        response = await client.callTool("get_project", { name: projName });
        break;
      }
      case "list_screens": {
        const rawScreensResp = await client.callTool("list_screens", { projectId: payload.projectId });
        const screensList = rawScreensResp?.screens || [];
        
        try {
          const projName = `projects/${payload.projectId}`;
          const projResp = await client.callTool("get_project", { name: projName });
          const instances = projResp?.project?.screenInstances || projResp?.screenInstances || [];
          
          const screenMap = new Map();
          screensList.forEach(s => {
            const sid = s.id || (s.name ? s.name.split('/').pop() : '');
            if (sid) screenMap.set(sid, s);
          });

          for (const inst of instances) {
            const sid = inst.id || (inst.sourceScreen ? inst.sourceScreen.split('/').pop() : '');
            if (sid && !screenMap.has(sid)) {
              screenMap.set(sid, {
                id: sid,
                name: inst.sourceScreen || `projects/${payload.projectId}/screens/${sid}`,
                title: inst.label || inst.title || `Tela ${sid.substring(0, 8)}`,
                description: inst.description || (inst.type === 'DESIGN_SYSTEM_INSTANCE' ? 'Instância de Design System' : ''),
                width: inst.width,
                height: inst.height,
                hidden: inst.hidden
              });
            } else if (sid && screenMap.has(sid) && inst.label) {
              screenMap.get(sid).title = inst.label;
            }
          }
          response = { screens: Array.from(screenMap.values()) };
        } catch (e) {
          response = rawScreensResp;
        }
        break;
      }
      case "upload_design_md": {
        response = await client.callTool("upload_design_md", payload);
        break;
      }
      case "create_design_system_from_design_md": {
        response = await client.callTool("create_design_system_from_design_md", payload);
        break;
      }
      default: {
        response = await client.callTool(command, payload);
        break;
      }
    }

    console.log(JSON.stringify(response, null, 2));
    try { await client.close(); } catch {}
  } catch (err) {
    console.error(JSON.stringify({
      error: err.message || String(err),
      details: err.stack
    }));
    process.exit(1);
  }
}

main();
